File: src/com/dafttech/terra/engine/renderer/TileRendererMultiblock.java
package com.dafttech.terra.engine.renderer;

import static com.dafttech.terra.resources.Options.BLOCK_SIZE;

import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.dafttech.terra.engine.AbstractScreen;
import com.dafttech.terra.engine.Vector2;
import com.dafttech.terra.game.world.Vector2i;
import com.dafttech.terra.game.world.World;
import com.dafttech.terra.game.world.entities.Entity;
import com.dafttech.terra.game.world.tiles.Tile;

public class TileRendererMultiblock extends TileRendererBlock {
    protected Vector2i multiblockSize;

    public TileRendererMultiblock(Vector2i multiblockSize) {
        this.multiblockSize = multiblockSize;
    }

    @Override
    public void draw(Vector2 pos, World world, AbstractScreen screen, Tile tile, Entity pointOfView, Object... rendererArguments) {
        Vector2 screenVec = tile.getPosition().toScreenPos(pointOfView);
        TextureRegion texture = tile.getImage();
        int cols = multiblockSize.x, rows = multiblockSize.y;
        int col = ((int) Math.abs(pos.x) % cols), row = ((int) Math.abs(pos.y) % rows);
        if (pos.x < 0) col = (cols - 1) - col;
        if (pos.y < 0) row = (rows - 1) - row;
        TextureRegion newTexture = new TextureRegion(texture);
        int width = newTexture.getRegionWidth() / cols, height = newTexture.getRegionHeight() / rows;
        int x = width * col, y = height * row;
        newTexture.setRegion(x, y, width, height);
        newTexture.flip(texture.isFlipX(), texture.isFlipY());
        screen.batch.draw(newTexture, screenVec.x + offset.x * BLOCK_SIZE, screenVec.y + offset.y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE);
    }
}

File: terra/src/main/scala/com/dafttech/terra/engine/renderer/TileRendererMultiblock.scala
package com.dafttech.terra.engine.renderer

import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.dafttech.terra.engine.TilePosition
import com.dafttech.terra.engine.{AbstractScreen, Vector2, Vector2i}
import com.dafttech.terra.game.world.World
import com.dafttech.terra.game.world.entities.Entity
import com.dafttech.terra.game.world.tiles.Tile
import com.dafttech.terra.resources.Options.BLOCK_SIZE
import scala.concurrent.duration._

object TileRendererMultiblock {
  def setRegion(texture: TextureRegion, x: Float, y: Float, width: Float, height: Float): Unit = {
    val invTexWidth: Float = 1.0F / texture.getTexture.getWidth
    val invTexHeight: Float = 1.0F / texture.getTexture.getHeight
    texture.setRegion(x * invTexWidth, y * invTexHeight, (x + width) * invTexWidth, (y + height) * invTexHeight)
  }
}

class TileRendererMultiblock extends TileRendererBlock {
  protected var multiblockSize: Vector2i = _

  def this(multiblockSize: Vector2i) {
    this()
    this.multiblockSize = multiblockSize
  }

  override def draw(screen: AbstractScreen, tile: Tile, pointOfView: Entity, rendererArguments: AnyRef*)(implicit tp: TilePosition): Unit = {
    val screenVec: Vector2 = tp.pos.toScreenPos(pointOfView)
    import monix.execution.Scheduler.Implicits.global
    val texture = tile.getImage.runSyncUnsafe(5.seconds)
    val cols: Int = multiblockSize.x
    val rows: Int = multiblockSize.y
    var col: Int = Math.abs(tp.pos.x).toInt % cols
    var row: Int = Math.abs(tp.pos.y).toInt % rows
    if (tp.pos.x < 0) col = ((cols - 1) - col + (cols - 1)) % cols
    if (tp.pos.y < 0) row = ((rows - 1) - row + (rows - 1)) % rows
    val newTexture: TextureRegion = new TextureRegion(texture)
    val width: Double = newTexture.getRegionWidth / cols.toDouble
    val height: Double = newTexture.getRegionHeight / rows.toDouble
    val x: Double = width * col
    val y: Double = height * row
    TileRendererMultiblock.setRegion(newTexture, x.toFloat, y.toFloat, width.toFloat, height.toFloat)
    newTexture.flip(texture.isFlipX, texture.isFlipY)
    screen.batch.draw(newTexture, (screenVec.x + offset.x * BLOCK_SIZE).toFloat, (screenVec.y + offset.y * BLOCK_SIZE).toFloat, BLOCK_SIZE, BLOCK_SIZE)
  }
}